### ISSUE TITLE
New Feature: Add a dedicated `Workspaces` page in settings 

### ISSUE THREAD

**Issue by trjExpensify** [Expensify/App#9971]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Deliverables** 

1. Add `Workspaces` as a menu item on the Avatar > Settings page, incl. a preview of 2 workspace avatars right-aligned on the row
2. Add a new page and move the list of workspaces to that page
3. Add a `New workspace` green footer button to the bottom of that new page
4. Remove `New workspace` from the three dots overflow menu that's accessed from a workspace (leave `Delete workspace` there) 

## Platform:

- To be implemented consistently across all platforms

**Notes/Photos/Videos:** 

![image](https://user-images.githubusercontent.com/16232057/179633661-f28298ba-f9e6-4733-8a72-1ca3f54c2f59.png)

[Figma file](https://www.figma.com/file/CL7LtJWt8Ax3GTEEMTYg6r/Workspace-Editor?node-id=301%3A1296)

**Expensify/Expensify Issue URL:** N/A 
**Issue reported by:** @trjExpensify 
**(Internal) Slack conversation:** https://expensify.slack.com/archives/C038T8LKKV5/p1657879133021399

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

CC: @shawnborton @JmillsExpensify as you may be interested in following along.

**Comment by trjExpensify** [comment]
I'll keep myself assigned for the CM portion. Passing this on to `engineering` first to triage.

**Comment by shawnborton** [comment]
Love it!

**Comment by thienlnam** [comment]
This will have to be on hold for API refactors (specifically for Policy_Create and Policy_Delete) cc @chiragsalian who is working on those -> currently held on Pattern B

After that, the rest of the items can be completed by an external contributor / or internal to make it quicker to correspond to design and UI items

Putting internal / on hold for now for these internal issues

https://github.com/Expensify/Expensify/issues/215184
https://github.com/Expensify/Expensify/issues/215181

**Comment by trjExpensify** [comment]
Ah cool, fair enough!

**Comment by thienlnam** [comment]
Still holding

**Comment by chiragsalian** [comment]
Discussed 1:1 with jack. Taking over the issue.

**Comment by chiragsalian** [comment]
The refactors all got unblocked and since those are the priority i'll first finish the API refactors and then get to this.

**Comment by trjExpensify** [comment]
Wahoo, nice! 😍

**Comment by JmillsExpensify** [comment]
Ya, seriously! This is a nice improvement.

**Comment by roryabraham** [comment]
If there's only one workspace, will we keep the same UI we have today?

**Comment by trjExpensify** [comment]
Nah, we'll move to the Workspaces page irrespective of how many you have. We chatted about that in the [OG thread](https://expensify.slack.com/archives/C038T8LKKV5/p1657879133021399), but to summarise: 

- it avoids the unpredictability of the UI moving around once you add more
- it makes it clear you can have multiple workspaces
- where to go to add more (with the addition of the `new workspace` button)

**Comment by trjExpensify** [comment]
@chiragsalian are we in a position to pick this one up now?

**Comment by JmillsExpensify** [comment]
Bumping that last question @chiragsalian

**Comment by chiragsalian** [comment]
Ah yes we are in a position to work on this now. Removing HOLD and setting to weekly.

**Comment by JmillsExpensify** [comment]
Thanks @chiragsalian! This is a nice win for our guides with tons of workspaces.

**Comment by trjExpensify** [comment]
💯 this is going to be super nice to declutter that top level settings page!

**Comment by chiragsalian** [comment]
I started working on this today and I'm halfway done with the functionality. I've to test it a little more on web and fix the bugs and then I'll test it on mobile too. Hoping to push out the PR today.

I had a question,
1. So from the screenshot, it shows some workspaces icons. Is that necessary and if a user has 10 workspaces are we going to just show the first two or will the UI be different.
![Screen Shot 2022-10-12 at 1 22 03 PM](https://user-images.githubusercontent.com/14356145/195440532-09161d2f-8c5d-4416-ab25-35ee6b5aa883.png)

cc @shawnborton if you have a suggestion.

**Comment by shawnborton** [comment]
I think we'd do something like we'll do for threads where maybe we show the first 3-4 avatars, and anything beyond that we just add a darker overlay with a `+X` counter on top to the last avatar:

<img width="264" alt="image" src="https://user-images.githubusercontent.com/2319350/195441251-80987356-ca6c-49ab-9095-ed0d0a9f7a5a.png">

**Comment by trjExpensify** [comment]
^^ I agree, love that. ❤️

**Comment by chiragsalian** [comment]
Question, how okay would it be if we reuse the same design we have for multiple chat participants in newDot, i.e., we currently show just one avatar followed by a number presented diagonally like so,
![image](https://user-images.githubusercontent.com/14356145/195468588-a01876f6-11bb-4368-b735-cbed48ef2be1.png)

If i re-use the same such that it looks like this on the workspace would that be fine?
![image](https://user-images.githubusercontent.com/14356145/195468673-6e78acc0-9054-44a0-9fdb-be74963954c3.png)

Or do you guys still prefer showing them stacked horizontally showing 3-4 avatars. Just wanted to confirm since if we go this route i would have to tweak the existing component. Alternatively, I'm also wondering if we push this out first since it's a quick improvement but then refine the design over time since that would take longer. Thoughts?

**Comment by chiragsalian** [comment]
Just a quick video showing the functionality. I've got a draft PR up. Most of the functionality is done minus just finalizing on the avatars. (see comment above regarding question on how to display the avatars)

In the video I also wanted to show how the avatars would work if there is an avatar set vs default since that was a bit tricky,

https://user-images.githubusercontent.com/14356145/195475019-1dbdf3e6-3e36-401d-ba36-46139805959e.mov

Things remaining - finalizing the avatars, testing on mobile, and some code cleanup before I can put this for review. From the video if you guys spot anything else that needs to change let me know.

**Comment by shawnborton** [comment]
I think we want to display the avatars like we have mocked up above and not the diagonal stack. I'm happy to create some exact mocks if what we need if that is helpful!

**Comment by chiragsalian** [comment]
Okay i think i've created a dynamic enough component to show avatars diagonally for chats and show stacked horizontally for. It looks like this for web,
![image](https://user-images.githubusercontent.com/14356145/195738478-7c0ef398-ac27-4a49-85e8-6ec30c52b8e8.png)

and like so on mobile,
![image](https://user-images.githubusercontent.com/14356145/195738525-732bf35f-ecbe-49af-add9-35d28c0b85c2.png)

Let me know what you guys think. My screen is pretty big so the content might look small, here is a zoomed one of how it looks on web,
![image](https://user-images.githubusercontent.com/14356145/195738724-1a0022a1-1e66-4ac4-aad7-9f2d147cf9e5.png)

If you prefer some size or position changes let me know or share a figma link, whichever is easier.

**Comment by shawnborton** [comment]
It looks like the icon on the left side of Workspaces is way too small now - I think you can just use the clean building.svg with no colored background in its place. 

Otherwise I think this is pretty good. The idea was that the `+X` number would show in a colored overlay on top of the last icon though, and I think we'd eventually reuse this pattern for threads too (like my mockup above has). Any reason not to take that approach now, or was it just easier to do it this way since the existing diagonal avatars look like this?

**Comment by chiragsalian** [comment]
> Any reason not to take that approach now, or was it just easier to do it this way since the existing diagonal avatars look like this?

Oh I can easily change the background or text color of the `+X`, can you share a mockup or maybe a color code for the background and text and i can update it.

**Comment by chiragsalian** [comment]
> It looks like the icon on the left side of Workspaces is way too small now - I think you can just use the clean building.svg with no colored background in its place.

Ah neat, i didn't notice in the mockup it was just the building. Updated it to be like so,
![image](https://user-images.githubusercontent.com/14356145/195899415-6cbb732d-9869-4dd9-be3d-cc0ec3130a6a.png)

![image](https://user-images.githubusercontent.com/14356145/195899555-a4b11278-3348-40b1-893b-2775b76aeabb.png)

**Comment by shawnborton** [comment]
@chiragsalian you can take our darkest brand navy color (#0B1B34) and use it at 64% opacity. Bonus points if we can make that into a color variable/reusable style. Then the number (+X) should just remain in white like it currently is.

**Comment by shawnborton** [comment]
Ah sorry - yes, the point is that it would show on top of an avatar behind it, like so:
<img width="868" alt="image" src="https://user-images.githubusercontent.com/2319350/195906257-2da452cf-5bfe-4712-a19e-a3070f21f099.png">

**Comment by chiragsalian** [comment]
oh that's different than the mockup in the issue description 💩 

oh, you want to show the 4rth avatar behind the `+X`. Interesting, okay I'll update it.  

There is another difference here vs the mockup and i wasn't sure which one to use. In the mockup the first avatar is behind the second one, while here it is inverse i.e., first avatar is above the second. Which one is the preference?

**Comment by shawnborton** [comment]
Good eye! I think it makes more sense to have "last on top" like this (my bad for the bad original screenies)
<img width="588" alt="image" src="https://user-images.githubusercontent.com/2319350/195944873-fec95617-c052-403d-968d-9f8e42b09823.png">

**Comment by chiragsalian** [comment]
Okay cool cool 👍 Thanks for the clarification

**Comment by chiragsalian** [comment]
The order is last on top. Btw yeah it makes more sense to have last on top so that the number doesn't get clipped. I made the number bolder so its easier to see.

Also added the avatar behind the dark opacity image. It looks like so now,
![image](https://user-images.githubusercontent.com/14356145/195949200-0d7e232d-9f60-40db-9ce9-604ccbc4afdf.png)

![image](https://user-images.githubusercontent.com/14356145/195949227-294a9aa5-389c-4c01-8f4f-8ca6d8406bcd.png)

Thoughts @shawnborton?

**Comment by shawnborton** [comment]
I think I would just leave the number in regular weight so it matches the numbers we use in the other avatar components. Otherwise it looks solid! I like that we can reuse this for threads too!

**Comment by chiragsalian** [comment]
Cool, ill leave the number with its regular weight. I think it's a little hard to read but we can tweak this as needed easily later too. Here is how it looks,
![image](https://user-images.githubusercontent.com/14356145/195952178-6186e448-cd44-489a-ad04-447da02a4b12.png)

![image](https://user-images.githubusercontent.com/14356145/195952211-3130781b-f277-4210-889c-d961b42782c4.png)

> I like that we can reuse this for threads too!

Yes, the plan is to have a reusable component for this. It may not work 100% for threads since some dimensions could change. But yeah we can always improve it further when we implement threads as needed.

**Comment by shawnborton** [comment]
Awesome, looks good - nice work!

**Comment by chiragsalian** [comment]
PR is almost ready, I've tested it on Web and iOS and fixed bugs/cleaned up the code. My android simulator isn't working so I'll get that sorted early next week and push the PR for review.

**Comment by chiragsalian** [comment]
Here is a short video to see its functionality in action, it shows how the UI looks when just 1, 2, 3, 4 and over 4 workspaces are present. 

https://user-images.githubusercontent.com/14356145/195959993-1e78b8c0-c9a8-4c57-a5a3-dc6a9dd0136a.mov

**Comment by shawnborton** [comment]
Just noting that we should add a more friendly "Empty state" to the blank workspaces page. I wonder if we should tackle that here? It would basically just be an Icon + text + button that feels more inviting than just a bunch of white space.

**Comment by shawnborton** [comment]
Otherwise this looks good to me! The counter logic is interesting though... I would think it would start off with `+1` instead of `+2`?

**Comment by chiragsalian** [comment]
> I wonder if we should tackle that here?

I'd prefer a follow-up if its okay. I don't mind working on it but my PR is already pretty large and it would be nice to tackle this separately as a smaller PR so its easier for the reviewers. But that's just my opinion.

> The counter logic is interesting though... I would think it would start off with +1 instead of +2?

Oh interesting. I just checked what slack does and it counts does not count the 4rth avatar in its `+x` (since it is shown in opacity). So it would show +1 instead of +2. Makes sense. My logic there was when the avatar wasn't shown in the background. Good catch. I'll update it.

**Comment by chiragsalian** [comment]
Cool yeah, i can easily add that in. @Expensify/marketing, could you suggest the copy for the middle of a blank workspace page i.e., this one?
![image](https://user-images.githubusercontent.com/14356145/196254748-001542ee-5190-4a68-97b1-bf553ec0b197.png)

**Comment by NickTooker** [comment]
@chiragsalian Thanks! I've read through the GH thread. Can you help me further understand what we are trying to convey to the end user with the text/copy specifically in the empty state? This will help me craft some versions for your review.

**Comment by NickTooker** [comment]
For the title - I'm not sure we need to change much from the title in the current mock.  I think this is exactly what we want the user to do. 

For the text - I am pulling from the help article that Shawn provided (which was great!).  I think we want to convey what the workspace does, and that it needs to be set-up in order to access it's features. Let me know you initial thoughts and we can tweak this together! 

**Create a Workspace**
_Workspaces are a one-stop shop for managing your settings, reimbursing receipts, managing cards, sending invoices, paying bills, and more! Create a new Workspace to access all available features!_

**Comment by shawnborton** [comment]
Nice, thanks Nick! What do you think about slightly trimming it down and omitting the `!`?

<img width="1434" alt="image" src="https://user-images.githubusercontent.com/2319350/196285193-0a408041-883c-4696-8df5-67f62609d73f.png">

**Comment by chiragsalian** [comment]
awesome, thanks guys, i'll update it.

**Comment by chiragsalian** [comment]
While working on the red brick road testing for another issue i just remembered that with the changes here whats best for showing the red dot. Currently on production its like this,
![image](https://user-images.githubusercontent.com/14356145/196305506-09bad7d1-40f5-4248-a426-a158fe89b29a.png)

 Should we just show the red dot on the left of the avatars here?
![image](https://user-images.githubusercontent.com/14356145/195952178-6186e448-cd44-489a-ad04-447da02a4b12.png)

Thoughts @shawnborton?

**Comment by shawnborton** [comment]
That makes sense to me - though i think I would put the red dot to the right of the avatars so that the red dot is always in the same spot near the arrow?

**Comment by chiragsalian** [comment]
Actually coming back to this, so shawn is it possible to provide mockups of how you want this to look with and without the red dot. Ideally, I would like things to be in the same place but I wanted to confirm the expectation. Will the avatars move a little left when there is a red dot and then move back more right when there is no red dot? Or were you thinking of static positions?

**Comment by shawnborton** [comment]
> Will the avatars move a little left when there is a red dot and then move back more right when there is no red dot? 

This is exactly what I was thinking. Happy to provide a mock if needed, but I think with flexbox you would just add the red dot inbetween avatars and the right arrow, all vertically middle-aligned.

**Comment by chiragsalian** [comment]
okay sounds good, i'll try that out.

**Comment by chiragsalian** [comment]
okay, how does this look. Can you check if the alignments are good. Two screenshots, one without redbrick and one with,

Web - Without red brick,
![image](https://user-images.githubusercontent.com/14356145/196804490-ba9793c7-2d7d-428a-b831-3cff24cd48b2.png)

Web - With red brick,
![image](https://user-images.githubusercontent.com/14356145/196802610-4e033730-ef18-4a8e-a2b9-ea5fd926cce6.png)

Mobile - Without Red brick
![image](https://user-images.githubusercontent.com/14356145/196804614-3b4c2b65-b5b1-4a24-97cd-2a14c3f0f112.png)

Mobile - With Red brick
![image](https://user-images.githubusercontent.com/14356145/196804333-110ca736-638e-41f6-9316-1b645444b88c.png)

Let me know if this looks good. The screenshot includes badge i.e., 0.00 next to Payments to confirm the alignment of workspaces and that was the same. I wasn't sure if it was, if i need to move it around let me know. Or if anything else should be adjusted.

**Comment by shawnborton** [comment]
I think that looks pretty good to me! Perhaps the badge/red dot could be 4px closer to the arrow though?

**Comment by jamesdeanexpensify** [comment]
I was looking back at the proposed copy for [this screen](https://github.com/Expensify/App/issues/9971#issuecomment-1281511248), and I'm wondering if the suggestion below is still accurate and a bit more succinct? Assuming mostly admins will create workspaces, at least to start...

> **Create a new Workspace**
> Workspaces are where you'll chat with your team, reimburse expenses, issue cards, send invoices, pay bills, and more -- all in one place.

**Comment by chiragsalian** [comment]
Sure, I moved the red dot 4px closer to the arrow. It looks like this,
![image](https://user-images.githubusercontent.com/14356145/196810694-1f3121b8-162f-431e-b14e-7d1f0adce21d.png)

Cool. I'm not sure of the copy so I'll let @NickTooker weigh in while i build that screen.

**Comment by shawnborton** [comment]
Sweet, that looks great to me - thanks Chirag!

**Comment by chiragsalian** [comment]
Shawn, does this look alright for the empty workspace page?
![image](https://user-images.githubusercontent.com/14356145/196819861-379a1ec2-e581-4215-a3c7-4ef654317ca1.png)

I feel like the subtitle is too close to the title, but i haven't changed it and just used the default values from the existing component. Let me know if you prefer I leave it like this or would prefer me to add some padding/margin after the title "Create a workspace".

Additionally, I'm waiting on the final copy before i ask for Spanish translation.

**Comment by shawnborton** [comment]
I agree with you, let's add maybe 4 or 8 more pixels in between title and body text to that main component.

**Comment by chiragsalian** [comment]
Done, this is margin-bottom 8,
Web -
![image](https://user-images.githubusercontent.com/14356145/196820789-c1fc49b3-a0df-44d6-9fb5-3fd4fdf84127.png)

Mobile - 
![image](https://user-images.githubusercontent.com/14356145/196821238-a939e948-3d3a-4eb4-9bc0-79fcf600c937.png)

I've updated the component vs having a different style just for this one. I think its fine since previous usages of the same component look fine to me,
![image](https://user-images.githubusercontent.com/14356145/196821042-f224577a-9049-4b94-8693-610f16d048a1.png)

**Comment by shawnborton** [comment]
Nice, that looks good to me.

**Comment by chiragsalian** [comment]
Moved the copy discussion to [slack](https://expensify.slack.com/archives/C03U7DCU4/p1666287655555599) to maybe finalize it with more speed.

**Comment by JmillsExpensify** [comment]
Woo! This looks awesome. Great work.

**Comment by chiragsalian** [comment]
Okay, I took some time fighting with the styles to get things consistent on Mobile/Web without any errors and testing took a bit of time too. Anyway, I finished testing, added screenshots, and just placed the code for review.

**Comment by trjExpensify** [comment]
Niceeee! Thanks, Chirag! ❤️

**Comment by aldo-expensify** [comment]
@chiragsalian  I think the red dot may be missing for the "Workspaces" entry in the settings menu:


https://user-images.githubusercontent.com/87341702/199127979-9e61842b-9350-4466-bdec-dc5158344d4c.mov


We may want to replace this line:

https://github.com/Expensify/App/blob/2f4bad7f4abf129daaa0e05f225245b076ac9f3c/src/libs/PolicyUtils.js#L51

by

```js
if (hasPolicyMemberError(policyMemberList) || hasCustomUnitsError(policy) || hasPolicyError(policy)) {
```

The detail with this fix is that it will make a red dot appear next to the workspace, and I think it shouldn't be there:

![image](https://user-images.githubusercontent.com/87341702/199128189-f05ee54f-a52b-45f0-97ce-49ab6a17af87.png)

so.. the better fix I think would be to replace the line here:

https://github.com/Expensify/App/blob/2f4bad7f4abf129daaa0e05f225245b076ac9f3c/src/pages/settings/InitialSettingsPage.js#L132

by

```js
.find(policy => PolicyUtils.getPolicyBrickRoadIndicatorStatus(policy, this.props.policyMembers) || PolicyUtils.hasPolicyError(policy))
```


![image](https://user-images.githubusercontent.com/87341702/199128267-303c4fa5-6d69-4d27-9c9b-3200dca3e90f.png)

**Comment by aldo-expensify** [comment]
Sorry, I think it is better if I just create a new GH issue for this and ping Chirag in case he wants to do it

**Comment by chiragsalian** [comment]
>  I think the red dot may be missing for the "Workspaces" entry in the settings menu:

That's odd. We tested the red dot for the Workspaces in the settings menu. I guess we must've missed a scenario since I tested it against members error and not an individual workspace error. Thanks for the catch.
